spike_multinvi.md

```diff
@@ -0,0 +1,205 @@
+
++// Helper
++private:
++   static bool SortPredicate(const OrbitalData& inLHS, const OrbitalData& inRHS);
++   void CalculateOrbitalDataMulti(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector);
++   void SortOrbitalVectorMulti(std::vector<OrbitalData>& ioOrbitalVector);
+
++// Data Members
++private:
++   std::size_t mNumThreads{0};
+};
+```
+
+Q: Could multithreading be achieved by making a new Multithreading class, as opposed to adding multithreading to a working single threaded class? 
+Q: Can we benefit from increased modularity and adds increased functionality without disrupting an already working implementation?
+Q: Is-a VS Has-a? Pros and cons of each? 
+Proposed:
+```
+// Single thread class
+class SatOrbitSingle
+{
+ // Interface
+public:
+    SatOrbit() = default;
+    virtual ~SatOrbit() = default;
+
+    std::vector<sat355::TLE> ReadFromFile(int argc, char* argv[]);
+    std::vector<OrbitalData> CalculateOrbitalData(const std::vector<sat355::TLE>& tleVector);
+    void SortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector);
+    std::vector<std::vector<OrbitalData>> CreateTrains(const std::vector<OrbitalData>& orbitalVector);
+    void PrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector);
+
+// Types
+protected:
+    using OrbitalDataVector = std::tuple<std::mutex, std::vector<OrbitalData>>;
+    using orbit_iterator = std::vector<OrbitalData>::iterator;
+
+// Implementation
+private:
+    // SatOrbit
+    virtual std::vector<sat355::TLE> OnReadFromFile(int argc, char* argv[]);
+    virtual void OnCalculateOrbitalData(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector);
+    virtual void OnSortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector);
+    virtual std::vector<std::vector<OrbitalData>> OnCreateTrains(const std::vector<OrbitalData>& orbitalVector);
+    virtual void OnPrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector);
+};
+
+
+// Multi thread class: Is-a SatOrbitSingle
+class SatOrbitMulti : public SatOrbitSingle
+{
+// Interface
+public:
+    SatOrbitMulti(std::size_t inNumThreads = 1);
+    ~SatOrbitMulti() override = default;
+
+// Types
+private:
+    using tle_const_iterator = std::vector<sat355::TLE>::const_iterator;
+    using IteratorPairVector = std::vector<std::tuple<orbit_iterator, orbit_iterator>>;
+
+// Implementation
+private:
+    // SatOrbitMulti
+    virtual void OnCalculateOrbitalDataMulti(const tle_const_iterator& tleBegin, const tle_const_iterator& tleEnd, OrbitalDataVector& ioDataVector);
+    virtual void OnSortOrbitalVectorMulti(orbit_iterator& inBegin, orbit_iterator& inEnd);
+    virtual void OnSortMergeVector(orbit_iterator& ioBegin, orbit_iterator& ioMid, orbit_iterator& ioEnd);
+
+// Data Members
+private:
+    std::size_t mNumThreads{0};
+};
+
+
+// Multi thread class: Has-a SatOrbitSingle
+class SatOrbitMulti
```

Comment:
I would recommend "3" classes here...
- i think we originally made a mistake with NVI, in that we started with "1" class that we expanded.
instead, i think we should have started with "2" classes: a) "pure abstract base `SatOrbit`", and b) "derived `SatOrbitImpl`"...

Now I recommend "3" classes:
- Pure abstract `SatOrbit`
- Derived `SatOrbitSingle`
- Derived `SatOrbitMulti`

The new `SatOrbitMulti` extends `SatOrbitSingle` only for those virtuals that need the multi-threaded treatment